File: 967-minimum-falling-path-sum/minimum-falling-path-sum.py
class Solution:
    def recursive(self, matrix, r, c):
        if c < 0 or c >= len(matrix[0]):
            return float('inf')
        if r == 0:
            return matrix[r][c]
        #Recursive
        left = float('inf')
        middle = float('inf')
        right = float('inf')
        left = self.recursive(matrix, r - 1, c - 1)
        middle = self.recursive(matrix, r - 1, c)
        right = self.recursive(matrix, r - 1, c + 1)
        result = matrix[r][c] + min(left, middle, right)
        return result

    def memoized(self, matrix, r, c, dp):
        if c < 0 or c >= len(matrix[0]):
            return float('inf')
        if r == 0:
            return matrix[r][c]
        if dp[r][c] != -1:
            return dp[r][c]
        #Recursive
        left = self.memoized(matrix, r - 1, c - 1, dp)
        middle = self.memoized(matrix, r - 1, c, dp)
        right = self.memoized(matrix, r - 1, c + 1, dp)
        result = matrix[r][c] + min(left, middle, right)
        dp[r][c] =  result
        return result
    
    def tabulation(self, matrix):
        dp = [[-1 for _ in range(len(matrix[0]))] for _ in range(len(matrix))]
        for r in range(len(matrix)):
            for c in range(len(matrix[0])):
                if r == 0:
                    dp[r][c] = matrix[r][c]
                else:
                    left = float('inf')
                    middle = float('inf')
                    right = float('inf')
                    if c - 1 >= 0:
                        left = matrix[r][c] + dp[r - 1][c - 1]
                    middle = matrix[r][c] + dp[r - 1][c]
                    if c + 1 < len(matrix[0]):
                        right = matrix[r][c] + dp[r - 1][c + 1]
                    dp[r][c] = min(left, middle, right)
        ans = float('inf')
        for val in dp[-1]:
            ans = min(ans , val)
        return ans

    def minFallingPathSum(self, matrix: List[List[int]]) -> int:
        return self.tabulation(matrix)

File: 967-minimum-falling-path-sum/minimum-falling-path-sum.cpp
class Solution {
public:
    // // TC -> O(N * M * 3 ^ N)
    // // SC -> O(N)
    // int recursive(vector<vector<int>>& matrix, int r, int c){
    //     if (c < 0 || c >= matrix[0].size()){
    //         return INT_MAX;
    //     }
    //     int row_size = matrix.size();
    //     int col_size = matrix[0].size();
    //     if (r == row_size - 1){
    //         return matrix[r][c];
    //     }
    //     int left = recursive(matrix, r + 1, c - 1);
    //     int right = recursive(matrix, r + 1, c + 1);
    //     int below = recursive(matrix, r + 1, c);
    //     int minimum = matrix[r][c] + min(min(left, right), below);
    //     return minimum;
    // }

    // // TC -> O(N * M * 3N)
    // // sc -> O(N * M + N)
    // int memoization(vector<vector<int>>& matrix,vector<vector<int>>& dp, int r, int c){
    //     if (c < 0 || c >= matrix[0].size()){
    //         return INT_MAX;
    //     }
    //     int row_size = matrix.size();
    //     int col_size = matrix[0].size();
    //     if (r == row_size - 1){
    //         return matrix[r][c];
    //     }
    //     if (dp[r][c] != -1){
    //         return dp[r][c];
    //     }
    //     int left = memoization(matrix, dp, r + 1, c - 1);
    //     int right = memoization(matrix, dp, r + 1, c + 1);
    //     int below = memoization(matrix, dp, r + 1, c);
    //     int minimum = matrix[r][c] + min(min(left, right), below);
    //     dp[r][c] = minimum;
    //     return minimum;
    // }

    int tabulation(vector<vector<int>>& matrix){
        int row_size = matrix.size();
        int col_size = matrix[0].size();
        vector<vector<int>> dp(row_size, vector<int>(col_size, -1));
        int ans;
        for (int r = row_size - 1; r >= 0; r--){
            ans = INT_MAX;
            for (int c = 0; c < col_size; c++){
                if (r == row_size - 1){
                    dp[r][c] = matrix[r][c];
                    ans = min(ans, dp[r][c]);
                    continue;
                } 

                int left = INT_MAX;
                int right = INT_MAX;
                int below = INT_MAX;

                if (c - 1 >= 0){
                    left = dp[r + 1][c - 1];
                }
                if (c + 1 < col_size){
                    right = dp[r + 1][c + 1];
                }
                below = dp[r + 1][c];
                dp[r][c] = matrix[r][c] + min(min(left, right), below);
                ans = min(ans, dp[r][c]);
            }
        }
        return ans;
    }
    int minFallingPathSum(vector<vector<int>>& matrix) {
        return tabulation(matrix);
    }
};